hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -1021,41 +1033,45 @@ void doDrain(final List<RAMQueueEntry> entries) throws InterruptedException {
 
   private void persistToFile() throws IOException {
     assert !cacheEnabled;
-    FileOutputStream fos = null;
-    ObjectOutputStream oos = null;
-    try {
+    try (ObjectOutputStream oos = new ObjectOutputStream(
+      new FileOutputStream(persistencePath, false))){
       if (!ioEngine.isPersistent()) {
         throw new IOException("Attempt to persist non-persistent cache mappings!");
       }
-      fos = new FileOutputStream(persistencePath, false);
-      oos = new ObjectOutputStream(fos);
+      oos.write(ProtobufUtil.PB_MAGIC);
+      byte[] checksum = PersistentIOEngineUtils.calculateChecksum(ioEngineName, algorithm);
```

Comment:
I get ur view on placing this checksum info.. The persist really comes into BucketCache as it persists meta info like backingMap.  Here the adjustment is keeping the ioEngineName as instance variable and the file paths are again derived from it.    In case of verify step that is very evident.  The readback from persistent file has to happen in BucketCache. And then the verification should be in FileIOE.  
How abt doing it this way?
We have isPersistent() in IOEngine.  Can we have an extended interface PersistentIOEngine which will have methods to get the checksum to persist and also for verify?  Within BC we can call these methods at appropriate places.

Reply:
In fact, I also thought about adding an interface PersistentIOEngine instead of PersistentIOEngineUtils. This is not done because other IOEngines (such as FileMmapIOE, although not yet in branch-1......) can reuse PersistentIOEngineUtils.

Reply:
I will submit a new version based on this idea